app/src/main/java/uk/gov/android/securestore/SharedPrefsStore.kt

```diff
@@ -1,42 +1,63 @@
 package uk.gov.android.securestore
 
 import android.content.Context
+import androidx.fragment.app.FragmentActivity
 import java.security.GeneralSecurityException
 import java.security.KeyStoreException
+import kotlin.coroutines.suspendCoroutine
+import uk.gov.android.securestore.authentication.AuthenticatorPromptConfiguration
 import uk.gov.android.securestore.crypto.CryptoManager
 import uk.gov.android.securestore.crypto.RsaCryptoManager
 
 class SharedPrefsStore(
-    context: Context,
+    context: FragmentActivity,
     configuration: SecureStorageConfiguration,
-    private val cryptoManager: CryptoManager = RsaCryptoManager()
+    private val cryptoManager: CryptoManager = RsaCryptoManager(
+        context,
+        configuration.id,
+        configuration.accessControlLevel
+    )
 ) : SecureStore {
-    private val sharedPrefs = configuration.getSharedPrefs(context)
+    private val sharedPrefs = context.getSharedPreferences(configuration.id, Context.MODE_PRIVATE)
 
-    override fun upsert(key: String, value: String): String {
-        try {
-            return cryptoManager.encryptText(key, value).also { writeToPrefs(key, it) }
-        } catch (e: GeneralSecurityException) {
-            throw SecureStorageError(e)
+    override suspend fun upsert(key: String, value: String): String {
+        return suspendCoroutine { continuation ->
+            try {
+                val result = cryptoManager.encryptText(value)
+                    .also { writeToPrefs(key, it) }
+                continuation.resumeWith(Result.success(result))
+            } catch (e: GeneralSecurityException) {
+                throw SecureStorageError(e)
+            }
         }
     }
 
     override fun delete(key: String) {
         writeToPrefs(key, null)
         try {
-            cryptoManager.deleteKey(key)
+            cryptoManager.deleteKey()
         } catch (e: KeyStoreException) {
             throw SecureStorageError(e)
         }
     }
 
-    override fun retrieve(key: String): String? {
-        try {
-            return sharedPrefs.getString(key, null)?.let {
-                cryptoManager.decryptText(key, it)
+    override suspend fun retrieve(
+        key: String,
+        authPromptConfig: AuthenticatorPromptConfiguration?
+    ): String? {
+        return suspendCoroutine { continuation ->
+            try {
+                sharedPrefs.getString(key, null)?.let { encryptedText ->
+                    cryptoManager.decryptText(
+                        encryptedText,
+                        { text -> continuation.resumeWith(Result.success(text)) },
+                        authPromptConfig
+                    )
+                    println("here3")
+                } ?: continuation.resumeWith(Result.success(null))
```

Comment:
Being unable to obtain a result is considered a successful result?

Reply:
It is potentially 'expected' behaviour given that it will return null if the desired key/value doesn't exist. So I wouldn't says it's a 'failure' if it doesn't find anything